Fix the issue that may create two campaigns after completing ads onboarding

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

While working on #2126, found another issue that may create two campaigns after completing ads onboarding. The root cause is [the reference changes of `onStatusApproved`](https://github.com/woocommerce/google-listings-and-ads/blob/60e1782084eb74ff6d7613ea3341136ea8ac3aae/js/src/components/paid-ads/billing-card/useAutoCheckBillingStatusEffect.js#L55) in `useAutoCheckBillingStatusEffect` hook, and it leads to calling `onStatusApproved` multiple times before redirecting out of the ads onboarding page.

This PR fixes it by ensuring the `onSetupComplete` callback of `BillingSetupCard` is only called once after getting an approved status for billing.

![1](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/45901eaf-6e01-4a54-a293-160a01b6506b)

### Screenshots:

https://github.com/woocommerce/google-listings-and-ads/assets/17420811/037ca89b-1f95-4f12-966c-5068e2e661c9

### Detailed test instructions:

1. `git checkout 3dd7efe8b8da3ed7d00b9ff17a559a5a43e0fb16`
2. `npm run env -- wp-scripts test-unit-js billing-setup-card` to see the failure of this test.
3. `git checkout ee841eba98df32ebc50d3661df529f06f3ea0d3c`
4. Rerun the above test to check if the test can pass.
5. `git checkout 3235e1496fc9a0bfd4ff7dc8ebd6506384935f9b`
6. Rerun the above test to see the failure of the second test case.
7. `git checkout fix/finish-onboarding-ads-duplicate-campaigns`
8. Rerun the above test to check if the second test case can pass.
9. Follow the reproduction steps in #2124 to see if it creates only a new campaign after finishing the ads onboarding flow 

### Changelog entry

> Fix - Avoid creating two campaigns after completing the Google Ads onboarding.

eason9487 committed:
Fix the issue that may create two campaigns after completing ads onboarding.

eason9487 committed:
Add a test case to test the number of calls for `onSetupComplete` callback of `BillingSetupCard`.

eason9487 committed:
Add another check to ensure the `useAutoCheckBillingStatusEffect` hook won't call back again by the next interval after billing is already approved.

eason9487 committed:
Add a test case to test if `onSetupComplete` of `BillingSetupCard` is not called again after the next interval.

jorgemd24 commented on js/src/components/paid-ads/billing-card/useAutoCheckBillingStatusEffect.js:
```diff
@@ -27,19 +27,29 @@ const completeGoogleAdsAccountSetup = () => {
 		} );
 
-		if ( billingStatus.status !== GOOGLE_ADS_BILLING_STATUS.APPROVED ) {
+		prevStatusRef.current = billingStatus.status;
```
Perhaps we could include a brief comment explaining why it's necessary to store the previous status. This would help us to undersand the code in the fure.

eason9487 replied:
Added in https://github.com/woocommerce/google-listings-and-ads/pull/2127/commits/93eeba40ca59a10fd200473c558758623c7629ed.

jorgemd24 commented on js/src/components/paid-ads/billing-card/billing-setup-card.test.js:
```diff
@@ -0,0 +1,114 @@
+			jest.runOnlyPendingTimers();
+		} );
+
+		expect( fetchBillingStatus ).toHaveBeenCalledTimes( 2 );
```
I would assume that once the billing status is approved, there's no need to call fetchBillingStatus again since it becomes unnecessary. WDYT?

eason9487 replied:
The `fetchBillingStatus` being called again is the internal implementation logic of `useAutoCheckBillingStatusEffect` hook, and yes, it's not needed for meeting the same business logic. However, the hook naming does not only scope the need for `checkStatusAndCompleteSetup` but also any possible changes of billing status, although there has been no that kind of need from the beginning till now.

From a test case perspective, it's not required to assert it. Removed by https://github.com/woocommerce/google-listings-and-ads/pull/2127/commits/f8e81083cee0b2c8dfacb0dea2652308cfe7a6f9.

jorgemd24 reviewed (approved):
Thanks @eason9487 for working on this. LGTM 👍 

I left a couple of comments but they are not blockers.

eason9487 committed:
Remove an unnecessary assertion from billing-setup-card.test.js

Address https://github.com/woocommerce/google-listings-and-ads/pull/2127#discussion_r1349564104

eason9487 committed:
Add a code comment for the status check condition in `useAutoCheckBillingStatusEffect`.

Address https://github.com/woocommerce/google-listings-and-ads/pull/2127#discussion_r1349561844
